lib/Web/Server.hs

```diff
@@ -42,6 +47,89 @@ url = BaseUrl Http "127.0.0.1" 8200 ""
 mount :: Text
 mount = "secret"
 
+data CofferServerError = CofferServerError
+  { cseError :: Text
+  , cseCode :: Int
+  }
+  deriving stock (Show, Generic)
+deriveJSON (aesonPrefix pascalCase) ''CofferServerError
+
+cofferServerErrorToBytes :: Text -> Int -> Bytes.Lazy.ByteString
+cofferServerErrorToBytes msg code =
+  CofferServerError msg code
+    & toEncoding
+    & encodingToLazyByteString
+
+throwViewError :: ViewResult -> Handler a
+throwViewError = \case
+  VRDirectory{} -> throwError err500 { errBody = "Expected view error, got VRDirectory." }
+  VREntry{} -> throwError err500 { errBody = "Expected view error, got VREntry." }
+  VRField{} -> throwError err500 { errBody = "Expected view error, got VRField." }
+  res@VRPathNotFound{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 100 }
+  res@VRDirectoryNoFieldMatch{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 101 }
+  res@VREntryNoFieldMatch{} -> throwError err404 { errBody = cofferServerErrorToBytes (pretty res) 102 }
+
+throwCreateError :: CreateResult -> Handler a
+throwCreateError = \case
+  CRSuccess{} -> throwError err500 { errBody = "Expected create error, got CRSuccess." }
+  CRCreateError createError ->
+    case createError of
+      res@CEParentDirectoryIsEntry{} -> throwError err409 { errBody = cofferServerErrorToBytes (pretty res) 200 }
+      res@CEDestinationIsDirectory{} -> throwError err409 { errBody = cofferServerErrorToBytes (pretty res) 201 }
+      res@CEEntryAlreadyExists{} -> throwError err409 { errBody = cofferServerErrorToBytes (pretty res) 202 }
+
+throwSetFieldError :: SetFieldResult -> Handler a
+throwSetFieldError = \case
+  SFRSuccess{} -> throwError err500 { errBody = "Expected set-field error, got SFRSuccess." }
```

Comment:
Don't get what you mean

Reply:
I mean, the only reason why this error 500 might happen is because we've separated the code like this:

```hs
public run token sfoQPath sfoFieldName  =
  run token ... >>= \case
      res@SFRSuccess{} -> pure res
      res@SFREntryNotFound{} -> throwSetFieldError res
      res@SFRMissingFieldContents{} -> throwSetFieldError res
```
```hs
throwSetFieldError :: SetFieldResult -> Handler a
throwSetFieldError = \case
  SFRSuccess{} -> throwError err500 ...
  res@SFREntryNotFound{} -> throwError $ addApplicationJSONHeader err404 ...
  res@SFRMissingFieldContents{} -> throwError $ addApplicationJSONHeader err400 ...
```

--------

There are two ways of eliminating this branch. One is by inlining `throwSetFieldError`:

```hs
public run token sfoQPath sfoFieldName  =
  run token ... >>= \case
      res@SFRSuccess{} -> pure res
      res@SFREntryNotFound{} -> throwError $ addApplicationJSONHeader err404
      res@SFRMissingFieldContents{} -> throwError $ addApplicationJSONHeader err404
```

Another way is by _not_ pattern matching on the `*Result` before calling `throwSetFieldError` (and also renaming it to simply `handleSetFieldResult`):

```hs
public run token sfoQPath sfoFieldName  =
  run token ... >>= handleSetFieldResult
```
```hs
handleSetFieldResult :: handleSetFieldResult -> Handler a
handleSetFieldResult = \case
  res@SFRSuccess{} -> pure res
  res@SFREntryNotFound{} -> throwError $ addApplicationJSONHeader err404 ...
  res@SFRMissingFieldContents{} -> throwError $ addApplicationJSONHeader err400 ...
```

Reply:
In the case of `throwSetFieldError`, it's used in the implementation of 3 endpoints, so option (2) is better.

But for most cases (like `throwDeleteFieldError` and `throwViewError`), they're only used in 1 endpoint each, so inlining them is better.

Reply:
Done

Reply:
> But for most cases (like `throwDeleteFieldError` and `throwViewError`), they're only used in 1 endpoint each, so inlining them is better.

I think you forgot this bit.

Reply:
Done